### PR TITLE
Add Quantize and Dequantize Linear op

### PR DESCRIPTION
**Pull request by kahmed10**

### CODE REVIEW

**Comment by kahmed10** [comment]
There should be no change to src/onnx/onnx.cpp, it looks like a glitch on github's end.

**Review by pfultz2 — APPROVED** [review]

File: src/onnx/parse_dequantizelinear.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            axis = info.attributes.at("axis").i();
+
+        auto input_lens = args[0]->get_shape().lens();
+
+        auto sub_zero_point = args[0];
+
+        if(args.size() == 3)
+        {
+            auto zero_point = args[2];
+            if(axis != 1)
```
Comment (scxiao) [inline src/onnx/parse_dequantizelinear.cpp:30]: Looks like the broadcasting here has some issues.
In my understanding, if it is a scalar shape, we can use multibroadcast; 
if it is a 1D tensor, it is a broadcast as make_op("broadcast", {{"axis", axis}, {"dims", input_lens}}) (not dependent on the axis value).

File: src/onnx/parse_dequantizelinear.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    instruction_ref parse(const op_desc& /*opd*/,
+                          const onnx_parser& /*parser*/,
+                          const onnx_parser::node_info& info,
+                          std::vector<instruction_ref> args) const
+    {
+        int axis = 1;
+        if(contains(info.attributes, "axis"))
+            axis = info.attributes.at("axis").i();
+
```
Comment (scxiao) [inline src/onnx/parse_dequantizelinear.cpp:22]: We should be able to handle the case of axis to be minus value as:
if (axis < 0) axis += n_rank;

File: src/onnx/parse_quantizelinear.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+struct parse_quantizelinear : op_parser<parse_quantizelinear>
+{
+    std::vector<op_desc> operators() const { return {{"QuantizeLinear"}}; }
+
+    instruction_ref parse(const op_desc& /*opd*/,
+                          const onnx_parser& /*parser*/,
+                          const onnx_parser::node_info& info,
+                          std::vector<instruction_ref> args) const
+    {
+        auto quant_type = shape::uint8_type;
```
Comment (scxiao) [inline src/onnx/parse_quantizelinear.cpp:19]: How do we know the default output type to be uint8_type?

File: test/onnx/gen_onnx.py [hunk 4]
```diff
@@ -1016,6 +1016,38 @@ def deconv_stride_test():
...
+    arg1 = helper.make_tensor_value_info('1', TensorProto.FLOAT, [5])
+    arg2 = helper.make_tensor_value_info('2', TensorProto.INT8, [5])
+    arg_out = helper.make_tensor_value_info('out', TensorProto.FLOAT,
+                                            [1, 1, 5, 1])
+
+    node = onnx.helper.make_node('DequantizeLinear',
+                                 inputs=['0', '1', '2'],
+                                 outputs=['out'],
+                                 axis=2)
+
```
Comment (scxiao) [inline test/onnx/gen_onnx.py:1047]: We can add a test case of the axis to be negative values.

**Review by scxiao — COMMENTED** [review]
You can turned on the onnx unit tests in the onnx_backend_test.py by adding the two lines:
        backend_test.include(r'.*test_dequantizelinear*')
        backend_test.include(r'.*test_quantizelinear*')

Some of the unit tests failed for now. They are supposed to pass if our implementation is correct.

File: src/onnx/parse_quantizelinear.cpp (hunk 3 above)
Comment (kahmed10) [reply src/onnx/parse_quantizelinear.cpp:19]: this is from the onnx spec which defaults the input to be uint_8

File: src/onnx/parse_dequantizelinear.cpp (hunk 1 above)
Comment (scxiao) [reply src/onnx/parse_dequantizelinear.cpp:30]: Another way for the broadcasting can be as:
For scalar shape, just do multibroadcasting, this is straightforward.
For the 1D case, we can first do a reshape as for the 1D tensor as {1, 1, (axis_dim use the dim of the 1D-tensor), 1), then do a multibroadcasting....

File: src/onnx/parse_dequantizelinear.cpp (hunk 2 above)
Comment (kahmed10) [reply src/onnx/parse_dequantizelinear.cpp:23]: fixed

File: test/onnx/gen_onnx.py (hunk 4 above)
Comment (kahmed10) [reply test/onnx/gen_onnx.py:1046]: added

**Review by scxiao — APPROVED** [review]

File: src/include/migraphx/tune_axis.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <utility>
+#include <cstdint>
+#include <string>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+inline int tune_axis(const int n_dim, const int axis, std::string op_name)
```
Comment (pfultz2) [inline src/include/migraphx/tune_axis.hpp]: Can we make `op_name` have default value?

Comment (kahmed10) [reply src/include/migraphx/tune_axis.hpp]: added

**Review by turneram — APPROVED** [review]

**Comment by mvermeulen** [comment]
Is there a reason some of the required items still seem to be pending?  Will merge once they pass.

**Comment by kahmed10** [comment]
Everything else was passing on the previous CI run, but for some reason github wasn't updating to reflect that. I just re-ran the job to see if it fixes the issue.
